internal/wasm/jit/jit_arm64.go

```diff
@@ -900,20 +906,171 @@ func (c *arm64Compiler) assignBranchTarget(labelKey string, br *obj.Prog) {
+	//
+	// Note: this is similar to how GCC implements Switch statements in C.
+	offsetData := make([]byte, 4*(len(o.Targets)+1))
+	c.addStaticData(offsetData)
+
+	// "tmpReg = &offsetData[0]"
+	c.compileConstToRegisterInstruction(
+		arm64.AMOVD,
+		int64(uintptr(unsafe.Pointer(&offsetData[0]))),
+		tmpReg,
+	)
+
+	// "index.register = tmpReg + (index.register << 2) (== &offsetData[offset])"
+	c.compileAddInstructionWithLeftShiftedRegister(index.register, 2, tmpReg, index.register)
+
+	// "index.regsetr = *index.reigier (== offsetData[offset])"
+	c.compileMemoryToRegisterInstruction(arm64.AMOVW, index.register, 0, index.register)
+
+	// Now we read the address of the beginning of the jump table.
+	// In the above example, this corresponds to reading the address of 0x123001.
+	c.compileReadInstructionAddress(obj.AJMP, tmpReg)
+
+	// Now we have the address of L0 in tmp register, and the offset to the target label in the index.register.
+	// So we could achieve the br_table jump by adding them and jump into the resulting address.
+	c.compileRegisterToRegisterInstruction(arm64.AADD, tmpReg, index.register)
+
+	c.compileUnconditionalBranchToAddressOnRegister(index.register)
+
+	// We no longer need the index's register, so mark it unused.
+	c.markRegisterUnused(index.register)
+
+	// [Emit the code for each targets and default branch]
+	labelInitialInstructions := make([]*obj.Prog, len(o.Targets)+1)
+	saved := c.locationStack
+	for i := range labelInitialInstructions {
+		// Emit the initial instruction of each target.
+		init := c.newProg()
+		// We use NOP as we don't yet know the next instruction in each label.
+		// Assembler would optimize out this NOP during code generation, so this is harmless.
+		init.As = obj.ANOP
+		labelInitialInstructions[i] = init
+		c.addInstruction(init)
+
+		var locationStack *valueLocationStack
+		var target *wazeroir.BranchTargetDrop
+		if i < len(o.Targets) {
+			target = o.Targets[i]
+			// Clone the location stack so the branch-specific code doesn't
+			// affect others.
+			locationStack = saved.clone()
+		} else {
+			target = o.Default
+			// If this is the deafult branch, we just use the original one
+			// as this is the last code in this block.
+			locationStack = saved
+		}
+		c.setLocationStack(locationStack)
+		if err := c.compileDropRange(target.ToDrop); err != nil {
+			return err
+		}
+		if err := c.compileBranchInto(target.Target); err != nil {
+			return err
+		}
+	}
+
+	// Set up the callbacks to do tasks which cannot be done at the compilation phase.
+	c.afterAssembleCallback = append(c.afterAssembleCallback, func(code []byte) error {
+		// Build the offset table for each target including default one.
+		base := labelInitialInstructions[0].Pc // This corresponds to the L0's address in the example.
+		for i, nop := range labelInitialInstructions {
+			if uint64(nop.Pc)-uint64(base) >= math.MaxUint32 {
```

Comment:
is there a way to fail earlier than here? or is it too indirect (to determine the potential pc length)

Reply:
unfortunately no way to fail as this is arch-specific (e.g. instruction length are different among architectures)